### PR TITLE
DPO Trainer

### LINKED ISSUE #405: Adding `DPOTrainer` in trl
DPO (Direct Preference Optimization) is a recent paper from Stanford University: https://arxiv.org/abs/2305.18290 
It could be interesting to implement a generalist trainer that supports this algorithm and provide an example script on how to use it

Screenshot 2023-06-05 at 18 22 15

Anyone on the community is free to start implementing it if they want to help on this! 

cc @natolambert @lewtun @lvwerra @TevenLeScao

### PR DESCRIPTION
**Pull request by kashif**
Initial DPOTrainer class for #405  by copying the ~~PPOTrainer~~ RewardTrainer and started to implement changes in it

Fixes #405

### CODE REVIEW

**Comment by Forbu** [comment]
Thank you @kashif.
I think one interesting stuff to do would be to make the forward pass of the model_ref in no_grad mode no ? 
You don't need to compute the backward pass for the ref model because you don't want to modify the ref model weights.

**Comment by kashif** [comment]
agree @Forbu and I think we will refactor the data-collator so that we only have a mask on the positive and negative part of the sequence...

**Comment by gaetanlop** [comment]
Hello @kashif, thanks for the DPO integration. I did a refactoring of the data collator to compute the mean logprobs only on the positive and negative part of the sequence. Can you share with me the branch?

**Comment by kashif** [comment]
@gaetanlop  i have added you to my fork

**Comment by gaetanlop** [comment]
Thanks @kashif, just pushed the required changes to make the trainer not compute the mean logprobs on masked input_ids. I used a similar approach as the HF DataCollatorForTokenClassification (https://github.com/huggingface/transformers/blob/main/src/transformers/data/data_collator.py)

**Comment by TevenLeScao** [comment]
Flagging that at some point we'll want the ScriptArguments to be consistent with the TrainingArguments (log_with -> report_to, batch_size -> per_device_train_batch_size, model_name -> model_name_or_path) the second and third ones are especially important since the semantic of the argument is actually different.

**Comment by TevenLeScao** [comment]
@kashif @gaetanlop a fix to enable distributed training required a change to `transformers.Trainer` (https://github.com/huggingface/transformers/pull/24326), you'll need to pull from master!

**Comment by kashif** [comment]
@TevenLeScao ok thanks!

**Comment by gaetanlop** [comment]
Hello @kashif @TevenLeScao, just wondering what misses from this PR in order to be merged in the main branch? If it misses experiments using the new DPO collator, I can try to do it

**Comment by kashif** [comment]
@gaetanlop yes lets try to do an experimental run and see, I believe that is what is missing

**Comment by kashif** [comment]
I will move this PR to "Ready for review"

File: examples/dpo.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from trl import DPOTrainer
+
+
+# Define and parse arguments.
+@dataclass
+class ScriptArguments:
+    """
+    The name of the Casual LM model we wish to fine with DPO
+    """
+
```
Comment (younesbelkada) [inline examples/dpo.py:26]:
```suggestion
```
nit - i don't think this is super informative

File: examples/dpo.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    max_length: Optional[int] = field(default=512, metadata={"help": "max length of each samples"})
+    label_pad_token_id: Optional[int] = field(default=-100, metadata={"help": "label for non response tokens"})
+    broadcast_buffers: Optional[bool] = field(default=False, metadata={"help": "ddp broadcast buffer argument"})
+
+
+parser = HfArgumentParser(ScriptArguments)
+script_args = parser.parse_args_into_dataclasses()[0]
+
+
+SANITY = True
```
Comment (younesbelkada) [inline examples/dpo.py:43]: To remove?

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+from ..core import logprobs_from_logits, masked_mean
+from ..import_utils import is_peft_available
+from .utils import DPODataCollatorWithPadding
+
+
+if is_peft_available():
+    from peft import get_peft_model
+
+
+def compute_accuracy(eval_pred) -> Dict[str, float]:
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:35]: I think this method is copied from: https://github.com/lvwerra/trl/blob/7705daa672f9264d7b5b789b6a1fd6b1cff03a58/trl/trainer/reward_trainer.py#L34 
Can you remove this method and move the method that is on the RewardTrainer to `trainer/utils` and import it from there?

File: trl/trainer/dpo_trainer.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    # We want to see how much of the time rewards_chosen > rewards_rejected.
+    predictions = np.argmax(predictions, axis=1)
+
+    accuracy = np.array(predictions == labels, dtype=float).mean().item()
+    return {"accuracy": accuracy}
+
+
+class DPOTrainer(Trainer):
+    r"""
+    The DPOTrainer.
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:47]: Maybe add few lines here and explain the list of arguments as done in the SFTTrainer: https://github.com/lvwerra/trl/blob/7705daa672f9264d7b5b789b6a1fd6b1cff03a58/trl/trainer/sft_trainer.py#L58 this will be useful to parse the documentation

File: trl/trainer/dpo_trainer.py [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+        callbacks: Optional[List[TrainerCallback]] = None,
+        optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (
+            None,
+            None,
+        ),
+        preprocess_logits_for_metrics: Optional[Callable[[torch.Tensor, torch.Tensor], torch.Tensor]] = None,
+        max_length: Optional[int] = None,
+        peft_config: Optional[Dict] = None,
+    ):
+        """
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:73]: I see you have done it here, I believe you can move it above similarly as : https://github.com/lvwerra/trl/blob/7705daa672f9264d7b5b789b6a1fd6b1cff03a58/trl/trainer/sft_trainer.py#L58

File: trl/trainer/dpo_trainer.py [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        Initialize DPOTrainer.
+
+        Args:
+            model (`transformers.PreTrainedModel`):
+                The model to train, preferably an `AutoModelForSequenceClassification`.
+            ref_model (`PreTrainedModelWrapper`):
+                Hugging Face transformer model with a casual language modelling head. Used for implicit reward computation and loss.
+            beta (`float`, defaults to 0.1):
+
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:82]: Description is missing

File: trl/trainer/dpo_trainer.py [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+                The maximum length of the sequences in the batch. This argument is required if you want to use the default data collator.
+            peft_config (`Dict`, defaults to `None`):
+                The PEFT configuration to use for training. If you pass a PEFT configuration, the model will be wrapped in a PEFT model.
+        """
+        if not is_peft_available() and peft_config is not None:
+            raise ValueError(
+                "PEFT is not installed and you passed a `peft_config` in the trainer's kwargs, please install it to use the PEFT models"
+            )
+        elif is_peft_available() and peft_config is not None:
+            model = get_peft_model(model, peft_config)
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:114]: Can you add a small check before the call to `get_peft_model` similarly as: https://github.com/lvwerra/trl/blob/main/trl/trainer/sft_trainer.py#L151-L152 to eventually prepare the quantized models? 🙏

File: trl/trainer/dpo_trainer.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+            eval_dataset,
+            tokenizer,
+            model_init,
+            compute_metrics,
+            callbacks,
+            optimizers,
+            preprocess_logits_for_metrics,
+        )
+
+        # Hack?
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:167]:
```suggestion
```

File: trl/trainer/dpo_trainer.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+        # Since we inherit from trainer we always have access to an accelerator
+        self.ref_model = self.accelerator.prepare_model(self.ref_model, evaluation_mode=True)
```
Comment (younesbelkada) [inline trl/trainer/dpo_trainer.py:169]: I would probably add a safety checker to check if `accelerator` is avaialble on the Trainer `hasattr(self, "accelerator")` - if not I would throw a warning to users and suggest them to update transformers - WDYT?

**Review by younesbelkada — COMMENTED** [review]
The integration looks awesome! Thanks a lot for working on this @kashif @TevenLeScao 
I left few minor comments, before merging, can you run the styling checks ? `make style && make quality` 
Also I think we should add a documentation section about DPOTrainer, I can help you on that, let me know

**Comment by TevenLeScao** [comment]
Addressed your issues @younesbelkada !

**Comment by vwxyzjn** [comment]
@kashif and @TevenLeScao really nice work! It appears a reference implementation from https://github.com/eric-mitchell/direct-preference-optimization just came up. It also has some nice wandb tracked experiments here https://wandb.ai/eric_anthony_mitchell/dpo-demos/table

image

Looks like the experiments only took 1 or 2 hours to complete, which seems quite doable 🙂. Should we include an example trying to match the learning curves?

**Comment by kashif** [comment]
yup will do!

**Comment by eric-mitchell** [comment]
Just to note: the few hours quoted are for 4 80G A100s, so runtimes will be longer if you use fewer devices. Excited to see DPO integrated into trl!

**Comment by eric-mitchell** [comment]
Awesome- if you do attempt to match our learning curves, happy to help out if any questions come up. You can reach out here, or on twitter or email.

**Comment by vwxyzjn** [comment]
Thanks @kashif and @eric-mitchell!. @kashif I updated the plotting utilities https://github.com/openrlbenchmark/openrlbenchmark/pull/25

```
pip install openrlbenchmark==0.2.1a4
python -m openrlbenchmark.rlops_multi_metrics \
    --filters '?we=eric_anthony_mitchell&wpn=dpo-demos&xaxis=_step&ceik=exp_name.value&cen=exp_name.value&metrics=rewards_train/chosen&metrics=rewards_train/accuracies&metrics=rewards_train/rejected' \
        'pythia28_hh_dpo_mp_bf16' \
    --filters '?we=costa-huang&wpn=trl&xaxis=_step&ceik=trl_ppo_trainer_config.value.tracker_project_name&cen=trl_ppo_trainer_config.value.log_with&metrics=env/reward_mean&metrics=objective/kl&metrics=objective/kl_coef' \
        'wandb?tag=gpt2-sentiment-1-nminibs&cl=TRL sentiment analysis (no minibatches, target kl=6)' \
    --env-ids pythia28_hh_dpo_mp_bf16 \
    --env-ids trl \
    --check-empty-runs \
    --pc.ncols 2 \
    --pc.ncols-legend 1 \
    --output-filename static/0compare 
```

creates 

image


We can probably use this to help with matching learning curves :)

**Review by younesbelkada — APPROVED** [review]
Great work! Thanks very much for your great effort @kashif @TevenLeScao and all the contributors! 
Leaving it to @lvwerra for the final words and I can address additional comments if he has any

File: trl/trainer/dpo_trainer.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+        with torch.no_grad():
+            logits_chosen_ref = self.ref_model(
+                input_ids=inputs["input_ids_chosen"],
+                attention_mask=inputs["attention_mask_chosen"],
+            )[0]
+            logits_rejected_ref = self.ref_model(
+                input_ids=inputs["input_ids_rejected"],
+                attention_mask=inputs["attention_mask_rejected"],
+            )[0]
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/dpo_trainer.py:188]: ah yes in the batch dim?

[the remark this answers is not in the document]
Comment (lvwerra) [reply trl/trainer/dpo_trainer.py:188]: yes exactly, what do you think?

[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/dpo_trainer.py:188]: yup makes sense i will do it!

**Comment by eric-mitchell** [comment]
@kashif @TevenLeScao @younesbelkada Once again, happy to help if anything weird shows up if you do end up doing a repro.

**Comment by kashif** [comment]
thanks @eric-mitchell  yes so we have the `dpo.py` script done and I can then try to adapt it to the datasets and tasks you used and compare once this PR is merged.

**Comment by lvwerra** [comment]
@kashif I think it would be great to run an experiment before we merge - otherwise we don't know if it works at all. What do you think?

**Comment by kashif** [comment]
@lvwerra sure! i can try then one of the ones from the official code repo as mentioned... i anyways wanted to do that

**Comment by eric-mitchell** [comment]
@kashif the easiest might be Pythia-2.8B on Anthropic-HH, since there are already demo wandb logs on the reference repo to compare with. I believe we did one epoch of SFT on chosen response and one epoch of DPO.

**Comment by kashif** [comment]
@eric-mitchell ok cool! I will start a script with that then! thanks!

**Comment by kashif** [comment]
@eric-mitchell I did an initial training with the Anthropic-hh dataset as per your original code and I am getting a much lower loss value than that in the wandb summary plots (around 0.3x). Would you have an idea what could be causing this? I Will debug on my side as well.

**Comment by eric-mitchell** [comment]
@kashif can you share your wandb plots/learning curves/rewards/etc.? How do the samples from your policy look?

We did fix a small tokenization bug in the DPO codebase since we generated the sample plots, and it's possible this affected the losses. I'll re-run our sample runs and see if the results are changed with this fix in.

**Comment by kashif** [comment]
@eric-mitchell  I have refactored my code and used the key etc. from your code now so I am testing that currently... will keep you posted!

**Comment by eric-mitchell** [comment]
@kashif just to follow up again, the re-run of the DPO stage of Pythia on HH just completed, and training is basically unchanged there as well after the bug fix on our end. So comparing with our original runs should be fine.

Screenshot 2023-07-05 at 2 36 21 PM


You can see the SFT/DPO runs before and after the bug fix [here](https://wandb.ai/eric_anthony_mitchell/dpo-demos).

File: trl/trainer/dpo_trainer.py [hunk 11]
```diff
@@ -0,0 +1,431 @@
...
+            A tensor of shape (batch_size,) containing the average/sum log probabilities of the given labels under the given logits.
+        """
+        assert logits.shape[:-1] == labels.shape
+
+        labels = labels[:, 1:].clone()
+        logits = logits[:, :-1, :]
+        loss_mask = labels != self.label_pad_token_id
+
+        # dummy token; we'll ignore the losses on these tokens later
+        labels[labels == self.label_pad_token_id] = 0
```
Comment (kashif) [inline trl/trainer/dpo_trainer.py:270]: @eric-mitchell  may I as if this `0` should be `self.padding_value` ?

Comment (eric-mitchell) [reply trl/trainer/dpo_trainer.py:270]: The value here doesn't matter as long as it's a valid token index (i.e. `[0, vocab_size)`). We're going to zero out the loss contribution for all of these tokens with `loss_mask` a few lines later. The reason we need to change it from -100 (which I assume is what the `label_pad_token_id` is) is that the gather will fail with negative indices.

Comment (kashif) [reply trl/trainer/dpo_trainer.py:270]: ah ok! got it!

**Comment by eric-mitchell** [comment]
@kashif just wanted to check in on this- curious if you've had the chance to re-run the replication experiment :)

**Comment by kashif** [comment]
@eric-mitchell i did but not with a SFT'ed pythia... also with Peft worked... We had validation loss around 0.6 or so as per your wandb so that was nice... if you can share your SFT'ed pythia model with me I can also run it now with that?

**Comment by eric-mitchell** [comment]
[Here](https://drive.google.com/file/d/1zp4H3YzADXx2eu_Y-AKVizOU6rT9sv2P/view?usp=sharing) are the weights to our pre-trained Pythia. You can load with `model.load_state_dict(torch.load(PATH)['state'])`.

It's nice that peft worked! What type of peft did you use?

**Comment by kashif** [comment]
@eric-mitchell thanks! yes we tried QLora and Lora as well.. let me confirm. Thanks for the weights!

File: docs/source/dpo_trainer.mdx [hunk 12]
```diff
@@ -0,0 +1,72 @@
+# DPO Trainer
```
Comment (younesbelkada) [inline docs/source/dpo_trainer.mdx:1]: I think you need to add that file in `_toctree.yml` file to see it on the documentation

Comment (kashif) [reply docs/source/dpo_trainer.mdx:1]: ah yes i fixed a conflict so will add it back

File: trl/trainer/dpo_trainer.py [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+
+        Args:
+            logits: Logits of the model (unnormalized). Shape: (batch_size, sequence_length, vocab_size)
+            labels: Labels for which to compute the log probabilities. Label tokens with a value of label_pad_token_id are ignored. Shape: (batch_size, sequence_length)
+            average_log_prob: If True, return the average log probability per (non-masked) token. Otherwise, return the sum of the log probabilities of the (non-masked) tokens.
+
+        Returns:
+            A tensor of shape (batch_size,) containing the average/sum log probabilities of the given labels under the given logits.
+        """
+        assert logits.shape[:-1] == labels.shape
```
Comment (lvwerra) [inline trl/trainer/dpo_trainer.py:263]: can we replace this with a proper error rather than assert?

File: docs/source/dpo_trainer.mdx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+TRL supports the DPO Trainer for training language models from preference data, as described in the paper [Direct Preference Optimization: Your Language Model is Secretly a Reward Model](https://arxiv.org/abs/2305.18290) by Rafailov et al., 2023.
```
Comment (lvwerra) [inline docs/source/dpo_trainer.mdx:3]: Let's link to the example script as well. E.g. "For a full example see .."

File: docs/source/trainer.mdx [hunk 15]
```diff
@@ -20,6 +20,10 @@ We also support a `RewardTrainer` that can be used to train a reward model.
 
 [[autodoc]] SFTTrainer
 
+## DPOTrainer
```
Comment (younesbelkada) [inline docs/source/trainer.mdx:23]: What do you think about adding one line in `index.mdx` linking to the doc page of `DPOTrainer` similarly as `SFTTrainer`, `RewardTrainer`, etc. ?

**Review by lvwerra — APPROVED** [review]
Two small nits, then we can merge :)
